### DISCUSSION TITLE
Store design leaks memory

### DISCUSSION THREAD

**Discussion by ostrolucky** [spatie/guzzle-rate-limiter-middleware#28] [General]
Looks like there wasn't any thought put into cleanup mechanism. This is especially visible when you look into built in InMemoryStore. So memory would usage would grow on every request, wouldn't it? If middleware was designed with this problem in mind, stores would either be aware of $timeframe so that they could clear everything older than it, or there would be some method for purging that could be called from RateLimiter. I wanted to use this middleware in persistent worker process so this isn't very good and will need to look elsewhere.

**Comment by AMDarter** [comment]
It seems this could create a scenario where my application is effectively in a perpetual sleep state. Are there any forks that address this issue?

**Comment by Tito1337** [comment]
The [example custom store](https://github.com/spatie/guzzle-rate-limiter-middleware?tab=readme-ov-file#custom-stores) is also a huge catastrophe, it never reduces the array nor expires the cache, leading to gigabytes of communication with Redis (or your other cache storage)
